docs/design.md

```diff
@@ -0,0 +1,785 @@
+# Deckhand Design
+
+## Purpose
+
+Deckhand is a document-based configuration storage service built with
+auditability and validation in mind.
+
+## Essential Functionality
+
+* layering - helps reduce duplication in configuration while maintaining
+  auditability across many sites
+* substitution - provides separation between secret data and other
+  configuration data, while allowing a simple interface for clients
+* revision history - improves auditability and enables services to provide
+  functional validation of a well-defined collection of documents that are
+  meant to operate together
+* validation - allows services to implement and register different kinds of
+  validations and report errors
+
+## Documents
+
+All configuration data is stored entirely as structured documents, for which
+schemas must be registered.
+
+### Document Format
+
+The document format is modeled loosely after Kubernetes practices. The top
+level of each document is a dictionary with 3 keys: `schema`, `metadata`, and
+`data`.
+
+* `schema` - Defines the name of the JSON schema to be used for validation.
+  Must have the form: `<namespace>/<kind>/<version>`, where the meaning of
+  each component is:
+  * `namespace` - Identifies the owner of this type of document. The
+    values `deckhand` and `metadata` are reserved for internal use.
+  * `kind` - Identifies a type of configuration resource in the namespace.
+  * `version` - Describe the version of this resource, e.g. "v1".
+* `metadata` - Defines details that Deckhand will inspect and understand. There
+  are multiple schemas for this section as discussed below. All the various
+  types of metadata include a `name` field which must be unique for each
+  document `schema`.
+* `data` - Data to be validated by the schema described by the `schema`
+  field. Deckhand only interacts with content here as instructed to do so by
+  the `metadata` section. The form of this section is considered to be
+  completely owned by the `namespace` in the `schema`.
+
+#### Document Metadata
+
+There are 3 supported kinds of document metadata. Documents with `Document`
+metadata are the most common, and are used for normal configuration data.
+Documents with `Control` metadata are used to customize the behavior of
+Deckhand. Documents with `Tombstone` metadata are used to delete pre-existing
+documents with either `Document` or `Control` metadata.
+
+##### schema: metadata/Document/v1
+
+This type of metadata allows the following metadata hierarchy:
+
+* `name` - string, required - Unique within a revision for a given `schema`.
+* `storagePolicy` - string, required - Either `cleartext` or `encrypted`. If
+  `encyrpted` is specified, then the `data` section of the document will be
```

Comment:
I think a per document specification makes a lot of sense but I really want to see what this looks like. Are we specifying roles? How is read and write of this particular kind for this particular role specified? What will the user experience be (auth error, empty returns (hidden elements), etc.)  What will the merge experience be with conflicting documentation authorization specs.  On the positive side, this approach means there is no implied security - its very explicit spelled out in the document.

To be sure, I think this approach can align with what I imagine is the goal here with deckhand the more we create very UCP-centric views the more it has to constantly adapt to new UCP elements/models in its API. If we instead are treating it as a generic repository it doesn't need to have a sprawling list of specific views with RBAC associated with them as the documents self-manage that.  That said, there is still no getting around spelling this out somehow -- that same UCP-centric specification just ends up requiring a thicker policy.yaml for deckhand with roles for each service articulated.

This also implies in some way that service-level authentication is occurring for the security to actually have any value -- initially we said that wasn't likely going to fit within the time frame and settled on Shipyard passing down its "long-lived" token throughout the chain.  Service accounts are of course more clean but we need to make sure we can implement it in a reasonable time frame.  In the end however, it doesn't matter either way to the deckhand framework itself - either we have a single shipyard role with very open permissions or we have achieved service level authentication where each service acts on behalf of itself and we can start closing things up into an as-needed-basis.